adapter/statsd/statsd.go

```diff
@@ -0,0 +1,161 @@
+// limitations under the License.
+
+package statsd
+
+import (
+	"bytes"
+	"fmt"
+	"text/template"
+	"time"
+
+	"github.com/cactus/go-statsd-client/statsd"
+	"github.com/golang/protobuf/ptypes/duration"
+
+	"istio.io/mixer/adapter/statsd/config"
+	"istio.io/mixer/pkg/adapter"
+)
+
+type (
+	// builder holds a connection to the statsd server which is shared by all aspects, as well as the parsed
+	// templates for constructing statsd metric names.
+	builder struct {
+		adapter.DefaultBuilder
+
+		client    statsd.Statter
+		templates map[string]*template.Template // metric name -> template
+	}
+
+	aspect struct {
+		rate      float32
+		client    statsd.Statter
+		templates map[string]*template.Template // metric name -> template
+	}
+)
+
+var (
+	name = "statsd"
+	desc = "Pushes statsd metrics"
+	// conf encodes the default configuration for this adapter
+	conf = &config.Params{
+		Address:                   "localhost:8125",
+		Prefix:                    "",
+		FlushInterval:             &duration.Duration{Seconds: 0, Nanos: int32(300 * time.Millisecond)},
+		FlushBytes:                512,
+		SamplingRate:              1.0,
+		MetricNameTemplateStrings: make(map[string]string),
+	}
+)
+
+// Register records the builders exposed by this adapter.
+func Register(r adapter.Registrar) {
+	r.RegisterMetricsBuilder(newBuilder())
+}
+
+func newBuilder() *builder {
+	client, err := statsd.NewBufferedClient(conf.Address, conf.Prefix, toTime(conf.FlushInterval), int(conf.FlushBytes))
+	if err != nil {
+		// TODO: something other than panic?
+		panic("failed to construct statsd client with err: " + err.Error())
+	}
+
+	templates := make(map[string]*template.Template)
+	for metricName, s := range conf.MetricNameTemplateStrings {
+		if t, err := template.New(metricName).Parse(s); err != nil {
+			// This should never happen given ValidateConfig, but we call this before validateconfig? unclear how this needs to work.
+			// We could migrate this parsing into a sync.Once and called during NewMetrics, but then it happen during a request.
+			panic(fmt.Errorf("failed to parse template '%s' for metric '%s' with err: %s", s, metricName, err))
+		} else {
+			templates[metricName] = t
+		}
+	}
+
+	return &builder{adapter.NewDefaultBuilder(name, desc, conf), client, templates}
+}
+
+func (b *builder) ValidateConfig(c adapter.AspectConfig) (ce *adapter.ConfigErrors) {
+	params, ok := c.(*config.Params)
+	if !ok {
+		// TODO: should this just report an error?
+		panic(fmt.Errorf("invalid config type, want adapter/statsd/config/config.proto, got: %v", params))
+	}
+
+	// TODO: we need to validate that the provided template strings are satisfiable by the mixer's attributes/metric's labels
+	for metricName, s := range params.MetricNameTemplateStrings {
+		if _, err := template.New(metricName).Parse(s); err != nil {
+			ce = ce.Append("template "+metricName, fmt.Errorf("failed to parse template '%s' for metric '%s' with err: %s", s, metricName, err))
```

Comment:
The first parameter to Append should just be the field name in the config where the error occurred. This will be used for error messages to the user.